cpp/src/parquet/row_range.cc

```diff
@@ -0,0 +1,225 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/row_range.h"
+
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/unreachable.h"
+#include "parquet/exception.h"
+
+namespace parquet {
+
+class IteratorImpl : public RowRanges::Iterator {
+ public:
+  explicit IteratorImpl(const RowRanges& ranges)
+      : iter_(ranges.ranges_.cbegin()), end_(ranges.ranges_.cend()) {}
+
+  ~IteratorImpl() override = default;
+
+  std::variant<RowRanges::IntervalRange, RowRanges::BitmapRange, RowRanges::End>
```

Comment:
I'm thinking that would variant being slow in storage? What would an internal:
```
enum class Type {
   IntervalRange,
   BitmapRange
};

struct Range {
   int64_t begin;
   /// for Interval this means range
   /// for bitmap this is a 64-bits map
   int64_t payload; 
};
std::vector<Type> types_;
std::vector<Range> ranges_;
```

I think this is more space efficient than current storage api...